pkg/proctree/proctree_feed.go

```diff
@@ -0,0 +1,257 @@
+package proctree
+
+import (
+	"github.com/aquasecurity/tracee/pkg/errfmt"
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/pkg/utils"
+)
+
+//
+// Feed process tree using signal events
+//
+
+type ForkFeed struct {
+	TimeStamp       uint64
+	ChildHash       uint32
+	ParentHash      uint32
+	LeaderHash      uint32
+	ParentTid       int32
+	ParentNsTid     int32
+	ParentPid       int32
+	ParentNsPid     int32
+	ParentStartTime uint64
+	LeaderTid       int32
+	LeaderNsTid     int32
+	LeaderPid       int32
+	LeaderNsPid     int32
+	LeaderStartTime uint64
+	ChildTid        int32
+	ChildNsTid      int32
+	ChildPid        int32
+	ChildNsPid      int32
+	ChildStartTime  uint64
+}
+
+// FeedFromFork feeds the process tree with a fork event.
+func (pt *ProcessTree) FeedFromFork(feed ForkFeed) error {
+	// DEBUG (TODO: remove this)
+	// file, _ := os.Open("/dev/null")
+	// file := os.Stdout
+	// fmt.Fprintf(file, "--\nFork event received:\n")
+	// fmt.Fprintf(file, "Parent Hash: %v\n", feed.ParentHash)
+	// fmt.Fprintf(file, "Leader Hash: %v\n", feed.LeaderHash)
+	// fmt.Fprintf(file, "Task   Hash: %v\n", feed.ChildHash)
+	// fmt.Fprintf(file, "PARENT:\ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.ParentTid, feed.ParentPid, feed.ParentNsPid, feed.ParentNsTid)
+	// fmt.Fprintf(file, "LEADER:\ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.LeaderTid, feed.LeaderPid, feed.LeaderNsPid, feed.LeaderNsTid)
+	// fmt.Fprintf(file, "CHILD: \ttid=%05d pid=%05d nspid=%05d nstid=%05d\n", feed.ChildTid, feed.ChildPid, feed.ChildNsPid, feed.ChildNsTid)
+	// END OF DEBUG
+
+	if feed.ChildHash == 0 || feed.ParentHash == 0 {
+		return errfmt.Errorf("invalid task hash")
+	}
+	if feed.ChildTid == 0 || feed.ChildPid == 0 {
+		return errfmt.Errorf("invalid child task")
+	}
+	if feed.ParentTid == 0 || feed.ParentPid == 0 {
+		return errfmt.Errorf("invalid parent task")
+	}
+
+	// Update the parent process (might already exist)
+
+	parent := pt.GetOrCreateProcessByHash(feed.ParentHash) // always a real process
+	parent.GetInfo().SetFeedAt(
+		TaskInfoFeed{
+			Tid:         int(feed.ParentTid),
+			Pid:         int(feed.ParentPid),
+			NsTid:       int(feed.ParentNsTid),
+			NsPid:       int(feed.ParentNsPid),
+			StartTimeNS: feed.ParentStartTime,
+			PPid:        -1, // do not change the parent PID
+			NsPPid:      -1, // do not change the ns parent PID
+			Uid:         -1, // do not change the parent UID
+			Gid:         -1, // do not change the parent GID
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
```

Comment:
Do we really need to update the parent if it already exists?

Reply:
Thats up to the strategy we want (https://github.com/aquasecurity/tracee/pull/3364#discussion_r1320779628). 

Decisions we can have:

1. Create process/thread at fork()
   - Update leader ? parent ? (they might have changed)
   - When to update process info ?
3. Access process tree node (process/thread) at any event
   (keep cache hot)
4. Update Binary/Comm at exec()
5. Update Exit at exit()

Some Ideas I had:

- To create a pipeline step close to "processEvent()":

![image](https://github.com/aquasecurity/tracee/assets/7395852/facf8a2a-3c95-40b1-ac73-2e5e71294c27)

- Each different event type would "enrich" the process tree IF enabled. Example: if you trace setuid() then each time its called it would enrich the process tree. Each time you "receive a DNS event" you would append the DNS domain as an artifact to the process in the process tree, etc.

Reply:
> Thats up to the strategy we want ([#3364 (comment)](https://github.com/aquasecurity/tracee/pull/3364#discussion_r1320779628)).
> 
> Decisions we can have:
> 
> 1. Create process/thread at fork()
>    
>    * Update leader ? parent ? (they might have changed)
>    * When to update process info ?

I don't think we should update leader not parent on fork. Just create the child. Anyway, other than comm (of the thread) what could have changed?

> 2. Access process tree node (process/thread) at any event
>    (keep cache hot)

Don't you do it in the current code?

> 3. Update Binary/Comm at exec()

Yes, we should do that (I think you already do it now?)

> 4. Update Exit at exit()
> 

Yes, we should do that (I think you already do it now?)

> Some Ideas I had:
> 
> * To create a pipeline step close to "processEvent()":
> 
> ![image](https://user-images.githubusercontent.com/7395852/266895211-facf8a2a-3c95-40b1-ac73-2e5e71294c27.png)
> 
> * Each different event type would "enrich" the process tree IF enabled. Example: if you trace setuid() then each time its called it would enrich the process tree. Each time you "receive a DNS event" you would append the DNS domain as an artifact to the process in the process tree, etc.

I'm not sure that adding a new stage to the pipeline is the best approach (performance wise). Let's discuss our options.

Reply:
We decided to add SignalProcessors (at the processEvent stage) instead of creating another stage in the pipeline (and I'll have a single handler for now, just to give directions for the next ones. Example: dealing with Setuid() when that event is enabled, for example.

Reply:
Not updating parent nor leader if they exist now.